### PR TITLE
 Create a class responsible for synchronizing file state.

### PR DESCRIPTION
**Pull request by rictic**
- [x] CHANGELOG.md not updated, internal change

---
This change is [Reviewable](https://reviewable.io/reviews/polymer/polymer-editor-service/79)

### CODE REVIEW

File: src/language-server/file-synchronizer.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    this._disposables.push(documents.onDidChangeContent((change) => {
+      // A document has changed in memory!
+      const workspacePath = converter.getWorkspacePathToFile(change.document);
+      this.inMemoryDocuments.set(workspacePath, change.document.getText());
+
+      fire([{type: FileChangeType.Changed, uri: change.document.uri}]);
+      // Publish document change so that we can clear our cache of the file.
+    }));
+
+    this._disposables.push(documents.onDidClose(async(event) => {
```
Comment (valdrinkoshi) [inline src/language-server/file-synchronizer.ts]: the callback doesn't need to be `async` here

File: src/language-server/file-synchronizer.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      // The file is no longer managed in memory, so we should delete it from
+      // the in-memory map.
+      this.inMemoryDocuments.delete(
+          converter.getWorkspacePathToFile(event.document));
+      fire([{type: FileChangeType.Changed, uri: event.document.uri}]);
+    }));
+
+    connection.onDidChangeWatchedFiles(async(req) => {
```
Comment (valdrinkoshi) [inline src/language-server/file-synchronizer.ts]: the callback doesn't need to be `async` here

File: src/language-server/util.ts [hunk 3]
```diff
@@ -17,7 +17,7 @@ import {Disposable} from 'vscode-languageserver';
 export class EventStream<T> {
   static create<T>() {
     const stream = new EventStream<T>();
-    const fire = stream.fire.bind(stream);
+    const fire: typeof stream.fire = stream.fire.bind(stream);
```
Comment (valdrinkoshi) [inline src/language-server/util.ts]: 🔮 👌

File: src/language-server/util.ts [hunk 4]
```diff
@@ -52,6 +52,7 @@ export class EventStream<T> {
   }
 }
 
+
```
Comment (valdrinkoshi) [inline src/language-server/util.ts]: nit: remove this extra space

File: src/language-server/util.ts (hunk 3 above)
Comment (rictic) [reply src/language-server/util.ts]: Yeah, hopefully someday typescript will handle `.bind()` reasonably.

File: src/language-server/file-synchronizer.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+    const {fire, stream} = EventStream.create<FileEvent[]>();
+    this.fileChanges = stream;
+    this._disposables.push(documents.onDidChangeContent((change) => {
+      // A document has changed in memory!
+      const workspacePath = converter.getWorkspacePathToFile(change.document);
+      this.inMemoryDocuments.set(workspacePath, change.document.getText());
+
+      fire([{type: FileChangeType.Changed, uri: change.document.uri}]);
+      // Publish document change so that we can clear our cache of the file.
```
[the remark this answers is not in the document]
Comment (rictic) [reply src/language-server/file-synchronizer.ts]: Moved and updated

**Pull request merged**